### ISSUE TITLE
Remove the `GET:audiences` endpoint and the `getAudiences()` selector, replacing usage of this with `getAvailableAudiences()`.

### ISSUE THREAD

**Issue by techanvil** [google/site-kit-wp#8487]
## Feature Description

As we're going to be caching the audiences in the DB, rather than retrieving them on-demand via the client, we will no longer need the `GET:audiences` endpoint and the corresponding `getAudiences()` selector.

These should be removed from the codebase. Existing use of `getAudiences()` should be replaced with the selector for retrieving the cached audiences, `getAvailableAudiences()`.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

- The `GET:audiences` endpoint and the `getAudiences()` selector should be removed from the codebase.
- Existing use of `getAudiences()` should be replaced with `getAvailableAudiences()`.
- The mapped, cached audience object for the `all-users` default audience should be updated so its `displayName` property is mapped to `'All visitors'`, this will fulfil the requirement to handle this display name as a special case (see [display names and descriptions](https://docs.google.com/document/d/1MGD5Djy6AeeZC4zBtHqS-lQEWD9jw0kf-IIWw-jLCFU/edit#heading=h.bvv6qc8gvz75) in the design doc).

## Implementation Brief

- [x] In `includes/Modules/Analytics_4.php`:
	- [x] Update the `get_datapoint_definitions` method:
		- [x] Remove the `GET:audiences` data point.
	- [x] Update the `create_data_request` method:
		- [x] Remove the case for the `GET:audiences` data point.
	- [x] Update the `set_available_audiences` method:
		- [x] When setting the `$audience_item` array, if the `displayName` in the received response is `All Users`, set it to `All visitors` in the `$audience_item` array.
- [x] In `assets/js/modules/analytics-4/datastore/audiences.js`:
	- [x] Remove the `getAudiences` selector and resolver.
	- [x] Remove the `fetchGetAudiencesStore` fetch store.
	- [x] Remove `reducerCallback` from the `fetchCreateAudienceStore` fetch store.
	- [x] Remove the `baseInitialState` object.
	- [x] Update the `createAudience` action:
		- [x] If the audience creation is successful, dispatch the `syncAvailableAudiences` action so that the list of audiences is updated in the `availableAudiences` module setting.
	- [x] Add a resolver for `getAvailableAudiences`:
		- [x] If the selector `getAvailableAudiences` returns `null`, dispatch the `syncAvailableAudiences` action and wait for it to complete.
- [x] After [the PR](https://github.com/google/site-kit-wp/pull/8497/) for [#8136](https://github.com/google/site-kit-wp/issues/8136) is merged:
	- [x] In `assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTiles.js`:
		- [ ] Replace the usage of the `MODULES_ANALYTICS_4` `getAudiences` selector with `getAvailableAudiences`.
	- [x] In `assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTiles.stories.js`:
		- [x] Remove usage of the `receiveGetAudiences` action, and use `setAvailableAudiences` instead with an appropriate fixture.
- [x] Replace any other usage of the `MODULES_ANALYTICS_4` `getAudiences` selector with `getAvailableAudiences`.

### Test Coverage

- [x] In `tests/phpunit/integration/Modules/Analytics_4Test.php`:
	- [x] Update the `test_get_datapoints__audienceSegmentation` test case:
		- [x] Remove the `audiences` string from the assertion array.
	- [x] Remove the `test_get_audiences` test case.
	- [x] Update the `test_sync_audiences` test case:
		- [x] Update the assertion for the `availableAudiences` module setting so that the first audience should have a `displayName` of `All visitors`.
- [x] In `assets/js/modules/analytics-4/datastore/audiences.test.js`:
	- [x] Remove test cases for the `getAudiences` selector.
	- [x] Update the `createAudience` `creates an audience` test case:
		- [x] Remove assertions regarding `store.getState().audiences`.
		- [x] Assert that it makes a `POST` fetch call to the `sync-audiences` endpoint.
		- [ ] If feasible, assert that it updates the `availableAudiences` module setting with the newly created audience.
- [x] Remove any fetch mock with the `/google-site-kit/v1/modules/analytics-4/data/audiences` endpoint.
- [x] Fix any other failing tests.

## QA Brief

* Make sure you have Site Kit set up with Google Analytics 4.
* Ensure the audienceSegmentation feature flag is enabled.
* Open the developer tools in your browser and navigate to the console tab.
* Execute the following command.

**Action: `getAvailableAudiences()`**

```js
googlesitekit.data.select('modules/analytics-4').getAvailableAudiences();
```

This should trigger a request to `google-site-kit/v1/modules/analytics-4/data/sync-audiences` endpoint and should give you the following response. 

This should give you the object in response.

```js
[
    {
        "name": "properties/437749618/audiences/8004992724",
        "displayName": "All visitors",
        "description": "All users",
        "audienceType": "DEFAULT_AUDIENCE",
        "audienceSlug": "all-users"
    },
    {
        "name": "properties/437749618/audiences/8004992725",
        "displayName": "Purchasers",
        "description": "Users who have made a purchase",
        "audienceType": "DEFAULT_AUDIENCE",
        "audienceSlug": "purchasers"
    }
]
```

Please note that this behaviour of calling the `sync-audiences` endpoint would occur only for the first time when `getAvailableAudiences` selector returns `null`. Afterwards, it will return the array of available audiences as stated above without calling the `sync-audiences` endpoint.

**AudienceTiles component check**

1. Review [AudienceTiles storybook component](https://google.github.io/site-kit-wp/storybook/develop/?path=/story/modules-analytics4-components-audiencesegmentation-audiencetiles--default) to ensure they align with the provided [Figma designs](https://www.figma.com/file/7pSrkEy8t00BcYRAi9LjjH/Audience-Segmentation?type=design&node-id=1046-39085&mode=design&t=P8twawG28pNjc0Lu-0) and AC.
2. Note that currently we are yet to have a functional tile component UI in the plugin, so should compare the storybook component.

## Changelog entry

* Replace direct calls to retrieve audiences with use of the cached list of audiences.

**Comment by kuasha420** [comment]
IB :heavy_check_mark:

**Comment by techanvil** [comment]
@nfmohit @kuasha420 it's worth bearing in mind that there will be additional use of `getAudiences()` by the time this issue is implemented (see the PR for its dependency 8136: https://github.com/google/site-kit-wp/pull/8497/files), so the IB and estimate for this should reflect that...

**Comment by kuasha420** [comment]
Ah, good point, I mixed up the ordering here, I thought 8136 was dependent on this one. Thanks @techanvil for catching it. 

@nfmohit quickly looking at the PR linked by Tom, it seems the selector can be drop-in replaced by `getAvailableAudiences` without much issue, but the tests need to be updated there as well. Can you take a closer look and update the IB accordingly?

**Comment by nfmohit** [comment]
On it, thank you @kuasha420 & @techanvil.

@techanvil: Do you think it'd be a good idea to add a resolver for `getAvailableAudiences` so that it dispatches the `syncAvailableAudiences` action when it is `null`?

**Comment by techanvil** [comment]
@nfmohit, that sounds like a good idea and one that would satisfy this requirement from the [audience caching](https://docs.google.com/document/d/1MGD5Djy6AeeZC4zBtHqS-lQEWD9jw0kf-IIWw-jLCFU/edit#heading=h.cwpr2gxebffj) spec in the design doc:

> - The list will be synced under the following circumstances:
>   - ...
>   - Any time the cached list is needed and the underlying setting doesn’t exist.

It _might_ turn out we sometimes want to retrieve the nullified, cached list without triggering a resync, in order to avoid making the request in some case, but this is not apparent at all at this stage, and it should be easy to find a solution to this if needed e.g. by adding a new selector. We can iterate our way across that bridge if the time comes, I think your suggestion makes sense in the meantime!

**Comment by nfmohit** [comment]
@kuasha420 I have updated the IB accordingly. I set a slightly higher estimate just to account for any other potential additions of the `getAudiences` selector. Let me know what you think, thanks!

**Comment by kuasha420** [comment]
@nfmohit  @techanvil Thank you both. I brought back the estimate one step back. (Not sure why ZenHub did it thrice) as the additional changes seems fairly straightforward. Cheers 

IB :heavy_check_mark:

**Comment by techanvil** [comment]
Hi @ankitrox, thanks for updating the PR and QAB. I've made a couple of minor tweaks to the PR to save the round trip. However, there's one last thing with regard to the QAB which I think it's worth sending back to you in order to help clarify the process.

We should generally be writing our QABs with QA engineers as the target audience (unless there's some aspect that does require a developer to perform some QA, but in this case it shouldn't). We don't expect QA engineers to have their own local dev environment set up, so we don't need to include instructions on building and viewing a local Storybook instance.

Please can you revise the QAB to remove this aspect and simply refer to the relevant stories. We often tend to provide links to the stories as they will appear on the deployed `develop` build, even though the links won't work until the PR is actually merged. For example you can take a look at the QAB for the recent issue https://github.com/google/site-kit-wp/issues/8230.

**Comment by kelvinballoo** [comment]
### QA Update ⚠️ 

Tested based on the QAB and results are as expected.
After executing the required command, the response matches similarly as per the QAB:
<img width="1297" alt="Screenshot 2024-05-01 at 16 37 46" src="https://github.com/google/site-kit-wp/assets/125576926/4e422326-ae04-403f-8648-3f1516de8e39">

The only point I wish to clarify from the QAB is on the need to:
**Review AudienceTiles storybook component to ensure they align with the provided Figma designs** 

This was already tested as part of https://github.com/google/site-kit-wp/issues/8136 previously.
I am assuming that maybe it was wrongly copied and pasted in this task because it feels fairly isolated from what we are achieving here.
Can you confirm @ankitrox ?

**Comment by kelvinballoo** [comment]
Based on latest conversations in Slack:
The changes made in the ticket changes some internals of AudienceTiles. So for a sanity check, we are just to ensure that the component is being loaded properly and there's no regressions. ie. missing data or console errors. 

There is no need to be as in-depth as in #8136. 
There was no style or other changes, so no visible difference should appear. 

The only change here should be the "All Users" tab has been changed to "All visitors" which is fine and expected.

To summarize, we are just to ensure that the story renders and there's no console errors or oddities.

**Comment by kelvinballoo** [comment]
### QA Update ✅ 

Reviewed the component and it's looking good.
'All Users' has been changed to 'All visitors' and there is nothing breaking on the different breakpoints.

<Details>
<img width="481" alt="Screenshot 2024-05-01 at 18 19 20" src="https://github.com/google/site-kit-wp/assets/125576926/b79bc1f1-60c8-4c24-99f8-7d0bf05aef04">

</Details>

Moving this ticket to Approval.
